### PR TITLE
jni: Add EnvoyEngine layer for library development

### PR DESCRIPTION
**Pull request by buildbreaker**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Add EnvoyEngine layer for library development
Risk Level: low
Testing: ci, local
Docs Changes: na
Release Notes: na
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: library/java/io/envoyproxy/envoymobile/EnvoyEngine.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    synchronized (EnvoyEngine.class) {
+      if (loader != null) {
+        return;
+      }
+
+      loader = new EnvoyEngine(context);
+    }
+  }
+
```
Comment (buildbreaker) [inline library/java/io/envoyproxy/envoymobile/EnvoyEngine.java]:
```suggestion
```

File: library/kotlin/io/envoyproxy/envoymobile/Envoy.kt [hunk 2]
```diff
@@ -1,25 +1,47 @@
...
 
-  private external fun runEnvoy(config: String): Int
-}
+  companion object {
+    @JvmStatic
+    fun load(context: Context) {
+      EnvoyEngine.load(context)
+    }
+  }
+}
```
Comment (buildbreaker) [inline library/kotlin/io/envoyproxy/envoymobile/Envoy.kt]:
```suggestion
}

```

File: library/java/io/envoyproxy/envoymobile/Envoy.java [hunk 3]
```diff
@@ -1,46 +1,14 @@
...
 
 import android.content.Context;
-import android.net.ConnectivityManager;
-import android.util.Log;
-
-import java.io.*;
+import io.envoyproxy.envoymobile.EnvoyEngine;
 
 // Wrapper class that allows for easy calling of Envoy's JNI interface in native Java.
 public class Envoy {
```
Comment (goaway) [inline library/java/io/envoyproxy/envoymobile/Envoy.java]: Are we going to port to Kotlin in a separate PR?

Comment (goaway) [reply library/java/io/envoyproxy/envoymobile/Envoy.java]: Nevermind, I see the implementation below.

File: BUILD [hunk 4]
```diff
@@ -53,8 +53,8 @@ kt_android_library(
 # This work around is to use an empty java_library to include the cc_library dependencies needed
 # for the kotlin jni layer
 android_library(
-    name = "java_cc_alias",
-    srcs = ["library/kotlin/io/envoyproxy/envoymobile/EnvoyKotlinEmptyClass.java"],
+    name = "envoy_engine",
```
Comment (goaway) [inline BUILD]:
```suggestion
    name = "envoy_engine_lib",
```
due to Bazel naming conventions.

File: BUILD [hunk 5]
```diff
@@ -31,7 +31,7 @@ touch $@
 
 aar_with_jni(
     name = "android_aar",
-    android_library = "android_lib",
+    android_library = "android_lib_kotlin",
```
Comment (goaway) [inline BUILD]: I think this can just stay the same (it's really only used internally anyways).

File: BUILD [hunk 6]
```diff
@@ -65,7 +65,7 @@ kt_android_library(
     srcs = ["library/kotlin/io/envoyproxy/envoymobile/Envoy.kt"],
     custom_package = "io.envoyproxy.envoymobile",
     manifest = "library/EnvoyManifest.xml",
-    deps = ["java_cc_alias"],
+    deps = ["envoy_engine"],
```
Comment (goaway) [inline BUILD]:
```suggestion
    deps = ["envoy_engine_lib"],
```

File: library/java/io/envoyproxy/envoymobile/Envoy.java (hunk 3 above)
Comment (buildbreaker) [reply library/java/io/envoyproxy/envoymobile/Envoy.java]: So the Kotlin version looks like this: https://github.com/lyft/envoy-mobile/pull/158/files#diff-6f7df3634fc91d9c4910cb9040b88705

I was going to delete the Java version after you got a chance to compare

File: library/kotlin/io/envoyproxy/envoymobile/Envoy.kt [hunk 7]
```diff
@@ -1,25 +1,47 @@
 package io.envoyproxy.envoymobile
 
 import android.content.Context
-import android.net.ConnectivityManager
+import io.envoyproxy.envoymobile.EnvoyEngine
 
-class Envoy {
+// Wrapper class that allows for easy calling of Envoy's JNI interface in native Java.
+class Envoy(
```
Comment (goaway) [inline library/kotlin/io/envoyproxy/envoymobile/Envoy.kt]: Kotlin looks so nice. :)

File: examples/java/hello_world/MainActivity.java [hunk 8]
```diff
@@ -19,8 +19,6 @@
 import java.util.List;
 import java.util.concurrent.TimeUnit;
 
-import io.envoyproxy.envoymobile.Envoy;
```
Comment (goaway) [inline examples/java/hello_world/MainActivity.java]: We don't still need this?

File: library/kotlin/io/envoyproxy/envoymobile/Envoy.kt (hunk 7 above)
Comment (buildbreaker) [reply library/kotlin/io/envoyproxy/envoymobile/Envoy.kt]: 🎉

**Pull request merged**